### ISSUE TITLE
The p-value of ologram change from one call to another

### ISSUE THREAD

**Issue by dputhier** [dputhier/pygtftk#87]
Th user wants to get something reliable. I guess the seed is not set somewhere. The problem was observed with the following command:

      gtftk join_attr -i mini_real.gtf.gz -H -j mini_real_counts_ENCFF630HEX.tsv -k gene_name -n exprs -t gene | gtftk discretize_key -k exprs -p -d exprs_class -n 6   | gtftk ologram -p ENCFF112BHN_H3K4me3_K562_sub.bed -c hg38.genome -D -n -m exprs_class -if example_pa_03_c.pdf -k 8 -j summed_bp_overlaps_pvalue

[example_pa_03_b.pdf](https://github.com/dputhier/pygtftk/files/3047987/example_pa_03_b.pdf)
[example_pa_03_a.pdf](https://github.com/dputhier/pygtftk/files/3047988/example_pa_03_a.pdf)
[example_pa_03_c.pdf](https://github.com/dputhier/pygtftk/files/3047989/example_pa_03_c.pdf)

**Comment by dputhier** [comment]
Changing itermax to 100000 in BetaCalculator does not fix the issue.

**Comment by dputhier** [comment]
But increasing itermax supress most of the warnings in the examples from the doc. So this is already a good point for #77 .

**Comment by dputhier** [comment]
This would also deserve a bats test to check the stability of the p-values.

**Comment by qferre** [comment]
Could you add print(x,a,b) to the part with the message in beta.py ? This would help me determine the range of values that cause this error.

It is very surprising, because this algorithm was not designed by me, and is supposed to be reliable...

**Comment by qferre** [comment]
I re-ran your command several times. The problem is not with the p-value calculation (although increasing the itermax will certainly help) : the shuffled esperances and variances were slighlty different, which means the shuffles were slightly different.

This is surprising, because I had not encountered this bug before even when running the same analysis several times.

There must be a random seed that is not set properly at some point, possibly due to a change we made. I am looking into this.

**Comment by qferre** [comment]
After doing some research, faulty random seed initialization is the most probable. I am looking into this.

This is especially worrisome since the bats test return the same value each time for some reason, as if the problem does not affect them.

**Comment by qferre** [comment]
My implementation of the random seed (only set it in ologram.py which is the main module) is the accepted way to do it.

Furthermore, when using ologram with fewer additional arguments, with this command :

```
gtftk ologram -p ENCFF112BHN_H3K4me3_K562_sub.bed -c hg38.genome -k 8
```

I get reproducible results with this !

So the problem must come from one of the additional parameters. Somehow.

**Comment by qferre** [comment]
I may have found the cause. If I remove the -n argument to process both basic features and "supplementary" features defined with -m, the results for the basic features are reproducible, but those for the supplementary features are not.

**Comment by qferre** [comment]
Using more_bed instead of more_features results in reproducible results, so the problem lies clearly with more_features.

**Comment by qferre** [comment]
*Found it*.

Here is the fautive line of code : ` user_key_values = set(gtf.extract_data(user_key))` followed by 
```
for val in user_key_values:
     do_something
```

@dputhier, @guillaumecharbonnier, and everyone working on this projet and all other projects : For the love of Zeus, please **never iterate on a Python set without sorting it before !** Sets are not ordered, hence the order of the iteration is not guaranteed, hence the shuffles would use a different sequence of random numbers, hence non reproducible results.

I now sort the set before processing it. This is fixed in f5e9eb4bb231466f5e567c78ae4543361f6799fc on the develop branch.
